Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
-        startColumn = startLoc.column
-      } else {
-        return nil
-      }
+    // Keep track if a line missing char should be printed.
+    var hasLineBeenSkipped = false
 
-      // Find the ending column.
-      let endLoc = highlight.endLocation(converter: slc, afterTrailingTrivia: false)
-      let endLine = endLoc.line
+    // Maximum numbers of digits in line number mark.
+    let maxNumberOfDigits = String(annotatedSourceLines.count).count
 
-      let endColumn: Int
-      if endLine > lineNumber {
-        endColumn = annotatedLine.sourceString.count
-      } else if endLine == lineNumber {
-        endColumn = endLoc.column
-      } else {
-        return nil
+    for annotatedLine in annotatedSourceLines {
+      /// If the current line number is not within the 'rangesToPrint', we continue to the next line
+      /// and set the 'hasLineBeenSkipped' flag, so that we can later insert a marker for the skipped lines.
+      guard rangesToPrint.contains(where: { $0.contains(annotatedLine.sourceLineNumber) }) else {
+        hasLineBeenSkipped = true
+        continue
       }
 
-      if startColumn == endColumn {
-        return nil
+      /// Add a skipped line marker to indicate that lines have been omitted, but only under two conditions:
+      /// 1. Lines have actually been skipped ('hasLineBeenSkipped' is true).
+      /// 2. There's already annotated content in 'annotatedSource', meaning this isn't the start of the annotated lines.
+      /// This ensures that skipped line markers only appear between annotated lines.
+      if hasLineBeenSkipped && !annotatedSource.isEmpty {
+        let skippedLineMarker = diagnosticDecorator.decorateBufferOutline("┆")
+        let skippedLine = indentString + String(repeating: " ", count: maxNumberOfDigits) + " " + skippedLineMarker + "\n"
+        annotatedSource.append(skippedLine)
       }
+      hasLineBeenSkipped = false
 
-      return startColumn..<endColumn
-    }.sorted { (lhs, rhs) in
-      lhs.lowerBound < rhs.lowerBound
-    }.mergingOverlappingRanges()
+      // Append a decorated current source line
+      appendDecoratedSourceLine(
+        to: &annotatedSource,
+        from: annotatedLine,
+        within: tree,
+        via: sourceLocationConverter,
+        indentingWith: indentString,
+        upToMaxDigits: maxNumberOfDigits
+      )
 
-    // Map the column ranges into index ranges within the source string itself.
-    let sourceStringUTF8 = annotatedLine.sourceString.utf8
-    let highlightIndexRanges: [Range<String.Index>] = highlightRanges.map { highlightRange in
-      let startIndex = sourceStringUTF8.index(sourceStringUTF8.startIndex, offsetBy: highlightRange.lowerBound - 1)
-      let endIndex = sourceStringUTF8.index(startIndex, offsetBy: highlightRange.count)
-      return startIndex..<endIndex
-    }
+      // If the line did not end with new line (e.g. the last line), append it manually
+      if annotatedSource.last?.isNewline == false {
```

Comment:
I generally find that my brain takes a lot more cycles to parse these `== true/false` comparisons on optionals than if using `??` to provide a default value for the `nil` case.

```suggestion
      if !(annotatedSource.last?.isNewline ?? false) {
```

Reply:
I understand the appeal of using `??` for clarity. However, in this case, I find `== true/false` clearer due to the double negation in `if !(annotatedSource.last?.isNewline ?? false) {`, which seems more complex. It feels simpler to directly compare the optional, avoiding the extra mental gymnastics.

Reply:
I'm quite convinced that using `== true/false` with optionals might seem a bit uncommon at first, but once you get accustomed to it, it becomes much simpler.

Reply:
If you think that your version is more readable, that’s a tie. Let’s leave your version then.